File: java/src/Day1.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;

public class Day1 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int sum1 = 0;
        int sum2 = 0;
        for (int i = 0; i < 1000; i++) {
            String s = br.readLine();
            sum1 += extract(s);
            sum2 += extract2(s);
        }
        System.out.println(sum1);
        System.out.println(sum2);
    }

    private static int extract(String s) {
        int first = -1;
        int last = -1;
        for (int i = 0; i < s.length(); i++) {
            if ('0' <= s.charAt(i) && s.charAt(i) <= '9') {
                first = s.charAt(i) - '0';
                break;
            }
        }
        for (int i = s.length() - 1; i >= 0; i--) {
            if ('0' <= s.charAt(i) && s.charAt(i) <= '9') {
                last = s.charAt(i) - '0';
                break;
            }
        }
        return first * 10 + last;
    }

    private static int extract2(String s) {
        Map<String, Integer> dict = new HashMap<>();
        dict.put("one", 1);
        dict.put("two", 2);
        dict.put("three", 3);
        dict.put("four", 4);
        dict.put("five", 5);
        dict.put("six", 6);
        dict.put("seven", 7);
        dict.put("eight", 8);
        dict.put("nine", 9);
        int firstIdx = Integer.MAX_VALUE;
        int firstVal = -1;
        int lastIdx = -1;
        int lastVal = -1;
        for (String key : dict.keySet()) {
            int idx = s.indexOf(key);
            if (idx >= 0) {
                if (idx < firstIdx) {
                    firstIdx = idx;
                    firstVal = dict.get(key);
                }
                if (s.lastIndexOf(key) > lastIdx) {
                    lastIdx = s.lastIndexOf(key);
                    lastVal = dict.get(key);
                }
            }
        }
        for (int i = 0; i < s.length(); i++) {
            if ('0' <= s.charAt(i) && s.charAt(i) <= '9' && i < firstIdx) {
                firstVal = s.charAt(i) - '0';
                break;
            }
        }
        for (int i = s.length() - 1; i >= 0; i--) {
            if ('0' <= s.charAt(i) && s.charAt(i) <= '9' && i > lastIdx) {
                lastVal = s.charAt(i) - '0';
                break;
            }
        }
        return firstVal * 10 + lastVal;
    }
}

File: python/day1.py
with open("input1.txt") as file:
    lines = [line.strip() for line in file.readlines()]
sum = 0
for line in lines:
    first = -1
    last = -1
    for char in line:
        if char.isdigit():
            if first == -1:
                first = int(char)
            last = int(char)
    sum += first * 10 + last
print(sum)

sum2 = 0
words = {"one": 1, "two": 2, "three": 3, "four": 4, "five": 5, "six": 6, "seven": 7, "eight": 8, "nine": 9}
for line in lines:
    firstIdx = 1e6
    lastIdx = -1
    firstVal = -1
    lastVal = -1
    for word in words:
        if word in line:
            if line.find(word) < firstIdx:
                firstIdx = line.find(word)
                firstVal = words[word]
            if line.rfind(word) > lastIdx:
                lastIdx = line.rfind(word)
                lastVal = words[word]
    for i in range(len(line)):
        if line[i].isdigit():
            if i < firstIdx:
                firstVal = int(line[i])
                break
    for i in range(len(line) - 1, -1, -1):
        if line[i].isdigit():
            if i > lastIdx:
                lastVal = int(line[i])
                break
    sum2 += firstVal * 10 + lastVal
print(sum2)